modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -401,19 +402,22 @@ private CompletableFuture<AsyncSqlCursor<List<Object>>> querySingle0(
                 .thenCompose(sqlNode -> {
                     final boolean rwOp = dataModificationOp(sqlNode);
                     final HybridTimestamp txTime = outerTx != null ? outerTx.readTimestamp() : rwOp ? null : clock.now();
+                    boolean distributedSqlDisabled = txTime == null;
+                    UUID txId = outerTx == null || distributedSqlDisabled ? null : outerTx.id();
 
                     BaseQueryContext ctx = BaseQueryContext.builder()
                             .frameworkConfig(
                                     Frameworks.newConfigBuilder(FRAMEWORK_CONFIG)
                                             .defaultSchema(schema)
-                                            .traitDefs(rwOp || (outerTx != null && !outerTx.isReadOnly()) ? Commons.LOCAL_TRAITS_SET :
-                                                    Commons.DISTRIBUTED_TRAITS_SET)
+                                            .traitDefs(distributedSqlDisabled ? Commons.LOCAL_TRAITS_SET : Commons.DISTRIBUTED_TRAITS_SET)
+                                            .traitDefs(Commons.DISTRIBUTED_TRAITS_SET)
```

Comment:
why do you have the row?  It looks as repeat )

Reply:
Thanks, fixed.